### PR TITLE
Do not use /dev/zero to mmap anonymous memory

### PR DESCRIPTION
**Pull request by VSadov**
Long time ago `/dev/zero` was commonly used to map anonymous memory because `MAP_ANON` / `MAP_ANONYMOUS` was not universally supported.

Nowdays there is no reason for that and using `MAP_ANON` / `MAP_ANONYMOUS` saves one fd per mapping.

The limit for fd count could be modest (1024 is a typical default) and these can add up quickly when loading assemblies from streams. 
Even when collectible context is used the count could be abnormally high since complete unloading requires 2 full GC cycles and can be delayed for other reasons.

Fixes:https://github.com/dotnet/runtime/issues/47859

### CODE REVIEW

File: src/coreclr/System.Private.CoreLib/src/System/Runtime/Loader/AssemblyLoadContext.CoreCLR.cs [hunk 1]
```diff
@@ -19,7 +19,7 @@ public partial class AssemblyLoadContext
 
         [RequiresUnreferencedCode("Types and members the loaded assembly depends on might be removed")]
         [DllImport(RuntimeHelpers.QCall, CharSet = CharSet.Unicode)]
-        private static extern IntPtr LoadFromStream(IntPtr ptrNativeAssemblyLoadContext, IntPtr ptrAssemblyArray, int iAssemblyArrayLen, IntPtr ptrSymbols, int iSymbolArrayLen, ObjectHandleOnStack retAssembly);
+        private static extern void LoadFromStream(IntPtr ptrNativeAssemblyLoadContext, IntPtr ptrAssemblyArray, int iAssemblyArrayLen, IntPtr ptrSymbols, int iSymbolArrayLen, ObjectHandleOnStack retAssembly);
```
Comment (VSadov) [inline src/coreclr/System.Private.CoreLib/src/System/Runtime/Loader/AssemblyLoadContext.CoreCLR.cs:22]: This looks like a typo that somehow makes no difference. The Qcall returns `void` -

https://github.com/dotnet/runtime/blob/69e114c1abf91241a0eeecf1ecceab4711b8aa62/src/coreclr/vm/assemblynative.cpp#L297

Comment (stephentoub) [reply src/coreclr/System.Private.CoreLib/src/System/Runtime/Loader/AssemblyLoadContext.CoreCLR.cs:22]: > that somehow makes no difference

Interesting.  Does it just end up returning some garbage register value or something?

**Review by stephentoub — APPROVED** [review]

File: src/coreclr/System.Private.CoreLib/src/System/Runtime/Loader/AssemblyLoadContext.CoreCLR.cs (hunk 1 above)
Comment (VSadov) [reply src/coreclr/System.Private.CoreLib/src/System/Runtime/Loader/AssemblyLoadContext.CoreCLR.cs:22]: Most likely it was returning in a register that caller immediately used to store something else. 

The original idea was that the mismatch in the returns was the reason for the "leak". Turns out it was a bit deeper than that, and fixing the return type did not change anything.

Comment (VSadov) [reply src/coreclr/System.Private.CoreLib/src/System/Runtime/Loader/AssemblyLoadContext.CoreCLR.cs:22]: cc: @janvorli

Comment (janvorli) [reply src/coreclr/System.Private.CoreLib/src/System/Runtime/Loader/AssemblyLoadContext.CoreCLR.cs:22]: The interop method would return whatever was in RAX / EAX / X0 / R0 at that point.

**Comment by VSadov** [comment]
Thanks!!

**Pull request merged**